src/metrics.jl

```diff
@@ -213,12 +213,12 @@ for dist in weightedmetrics
     @eval parameters(d::$dist) = d.weights
 end
 
-result_type(dist::UnionMetrics, a::AbstractArray, b::AbstractArray) =
-    result_type(dist, a, b, parameters(dist))
-result_type(dist::UnionMetrics, a::AbstractArray, b::AbstractArray, ::Nothing) =
-    typeof(_evaluate(dist, oneunit(eltype(a)), oneunit(eltype(b))))
-result_type(dist::UnionMetrics, a::AbstractArray, b::AbstractArray, p) =
-    typeof(_evaluate(dist, oneunit(eltype(a)), oneunit(eltype(b)), oneunit(eltype(p))))
+result_type(dist::UnionMetrics, a::AbstractArray, b::AbstractArray, p=nothing) =
+    result_type(dist, eltype(a), eltype(b), isnothing(p) ? parameters(dist) : p)
```

Comment:
It seems there is no need for `isnothing` here if you just define
```suggestion
result_type(dist::UnionMetrics, a::AbstractArray, b::AbstractArray, p=parameters(dist)) =
    result_type(dist, eltype(a), eltype(b), p)
```
BTW depending on your use case in ImageDistances (which is not completely clear to me) it might be sufficient to define
```suggestion
result_type(dist::UnionMetrics, ::Type{Ta}, ::Type{Tb}) where {Ta,Tb} =
    result_type(dist, Ta, Tb, parameters(dist))
```
since the fallback is `result_type(dist::PreMetric, a::AbstractArray, b::AbstractArray) = result_type(dist, eltype(a), eltype(b))`.

Reply:
Indeed, the second suggestion works just fine (without the compat stuff), after you remove the `p` from the arguments here:

https://github.com/JuliaStats/Distances.jl/blob/d547d3a78db98411a902fe5382243ece00bfdcc3/src/metrics.jl#L264